osu.Game/Screens/Backgrounds/EditorBackgroundScreen.cs

```diff
@@ -0,0 +1,117 @@
+    {
+        private readonly WorkingBeatmap beatmap;
+        private readonly Container dimContainer;
+
+        private CancellationTokenSource? cancellationTokenSource;
+        private Bindable<float> dimLevel = null!;
+        private Bindable<bool> showStoryboard = null!;
+
+        private BeatmapBackground background = null!;
+        private Container storyboardContainer = null!;
+
+        private IFrameBasedClock? clockSource;
+
+        public EditorBackgroundScreen(WorkingBeatmap beatmap)
+        {
+            this.beatmap = beatmap;
+
+            InternalChild = dimContainer = new Container
+            {
+                RelativeSizeAxes = Axes.Both,
+            };
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config)
+        {
+            dimContainer.AddRange(createContent());
+            background = dimContainer.OfType<BeatmapBackground>().Single();
+            storyboardContainer = dimContainer.OfType<Container>().Single();
+
+            dimLevel = config.GetBindable<float>(OsuSetting.EditorDim);
+            showStoryboard = config.GetBindable<bool>(OsuSetting.EditorShowStoryboard);
+        }
+
+        private IEnumerable<Drawable> createContent() =>
+        [
+            new BeatmapBackground(beatmap) { RelativeSizeAxes = Axes.Both, },
+            // this kooky container nesting is here because the storyboard needs a custom clock
+            // but also needs it on an isolated-enough level that doesn't break screen stack expiry logic (which happens if the clock was put on `this`),
+            // or doesn't make it literally impossible to fade the storyboard in/out in real time (which happens if the fade transforms were to be applied directly to the storyboard).
+            new Container
+            {
+                RelativeSizeAxes = Axes.Both,
+                Child = new DrawableStoryboard(beatmap.Storyboard)
+                {
+                    Clock = clockSource ?? Clock,
+                }
+            }
+        ];
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            dimLevel.BindValueChanged(_ => dimContainer.FadeColour(OsuColour.Gray(1 - dimLevel.Value), 500, Easing.OutQuint), true);
+            showStoryboard.BindValueChanged(_ => updateState());
+            updateState(0);
+        }
+
+        private void updateState(double duration = 500)
+        {
+            storyboardContainer.FadeTo(showStoryboard.Value ? 1 : 0, duration, Easing.OutQuint);
+            // yes, this causes overdraw, but is also a (crude) fix for bad-looking transitions on screen entry
+            // caused by the previous background on the background stack poking out from under this one and then instantly fading out
+            background.FadeColour(beatmap.Storyboard.ReplacesBackground && showStoryboard.Value ? Colour4.Black : Colour4.White, duration, Easing.OutQuint);
```

Comment:
What's the reasoning for this being white when there's no storyboard present?

Reply:
The part that's probably tricky here is that it's not straight white per se, it's *multiplicative white* applied to the *beatmap background*, i.e. just unchanged beatmap background colours.

The goal was to have the background double for both a black box when storyboard is showing and it replaces the background, to ensure nothing drawn behind pokes out, and a normal background in every other circumstance.

Reply:
Ah right, makes more sense now.